### DISCUSSION TITLE
MAX_SWITCHES > 8 on ESP32?

### DISCUSSION THREAD

**Discussion by jig432** [arendst/Tasmota#10373] [Ideas]
I'd like to be able to use an esp32 to monitor more than 8 (switch/button) inputs. Specifically I'd like 10, but it has plenty of pins that could be used as switches.  I have a security system I want to monitor 10 zones.  So far I am using a custom arduino image, but would like to gain all the advantages of using tasmota.
I naively tried to change MAX_SWITCHES to 10 and compile and that didn't boot.  I browsed the code some, and saw an enum with entries for SET_SWITCH_TXTx (x=1-8) that likely would need to be adjusted.  But in a quick browse didn't see anything else that jumped out as a problem. I suspect there are assumptions for at  most 8 switches.  I'm hoping someone more familiar with the code knows what the problem might be or where to look.  Is this something that could be fairly easily fixed?  Maybe with a new define to allow private compiles without burdening the standard images.

**Comment by barbudor** [comment]
Hi @jig432 
Increasing the number of switches will impact multiple elements and more specifically the Settings section that is stored in flash.
Changing the define right now would break the structure of the flash and makes it non interoperable with previous version.
So requires a little double-thinking...

**Reply by jig432** [reply]
I suspected such a change could impact flash structure and have compatibility issues.  That's partly why I wondered if a new define for private compiles could work.  That way anyone wanting such support could be warned about incompatibilities.  Of course, depending on how severe the incompatibilities turned out to be. Any thinking would be much appreciated. Also, if there is anything I can do to help, let me know.

**Comment by gemu2015** [comment]
with scripts you may use as many switches a the chip has gpios

**Reply by jig432** [reply]
I'm still learning tasmota. Scripts sound good.  I built an esp32 version (based on  9.1.0) of tasmota with #define USE_SCRIPT (and undefine USE_RULES).  And have what I hope is the start of a script for gpio pins.
However, I have run into an issue with getting parameter substitution working in a subroutine.  I tried various attempts at parameters (patterned after the scripting doc examples):
```
>D
t="hi"
>BS
=#sub1("hallo")
=#sub1(t)
#sub1(string)
print a string
print b %string%
=>print c %string%

```
and get this as output:
```
00:40:42 a string                                                                                                       
00:40:42 b ???                                                                                                          
00:40:42  c ???                                                                                                         
00:40:42 a string                                                                                                       
00:40:42 b ???                                                                                                          
00:40:42  c ??? 
```

Is there some other #define needed to make subroutine parameters work? Or did I do something else wrong? Also, not sure why there is an extra leading " " for the "=>print" case.  A "print %t%" in the BS section works as expected, so basic variable substitution works.

**Reply by jig432** [reply]
followup/question: I found that if I declare the parameter in the >D section, it works.  But is only 1 parameter supported for a subroutine?
Doing this:
```
>D
nam="a"
uid="c"

>B
print hi
=#pubcfg("hallo","b")

#pubcfg(nam,uid)
=>print %nam%-%uid%

```
produces this output:
```
1:22:37 hi                                                                                                             
01:22:37  a-c  
```

**Comment by gemu2015** [comment]
yes only 1 parameter supported for subroutines
and there are no local variables, so you have to declare the parameter in >D
and delimiter between parameters of functions is always the space character not comma

i am aware that the documentation often lacks clarity
everybody is invited to help improving the docs, just edit  the source and submit a pr

**Comment by gemu2015** [comment]
@jig432 
Thank you for your enhancement of scripter in source and docs!
Keep going!

**Reply by jig432** [reply]
Glad to help.  I had hoped adding multiple params for subroutines would be straightforward, and it was.  Also, just submitted a pr to docs to update and expand description of subroutines.  Hope that is clear and correct and will help others.

**Comment by jig432** [comment]
FYI, I have 10 gpios working, publishing mqtt status changes and HA config entries, via a script.  Thanks for scripting support and for your help answering questions!

**Reply by bartek666666** [reply]
could you share your script?

**Reply by jig432** [reply]
Here is most of the script I use for 10 GPIOs.  Probably lots of room for improvement. I removed 8 additional pubcfg calls below.  But this should give you the idea. Since I am using the GPIOs to monitor a security system, I also used my own mqtt topics.


\>D
maxp=10
m:pins=0 10
m:pst=0 10
p=1
zb=8
zn=0
cfg=0
i=0
wind="window"
nam=""
zstr=""
cls=""
uid=""
mm=0

\>B
pins[1]=15
pins[2]=4
pins[3]=16
pins[4]=17
pins[5]=5
pins[6]=18
pins[7]=19
pins[8]=21
pins[9]=22
pins[10]=23
for i 1 maxp 1
p=pins[i]
spinm(p 0)
if pin[p]==0
then
pst[i]=1
else
pst[i]=0
endif
next

\>S <execute once per second>
if mqtts>0
then
mm=0
; check switches
for i 1 maxp 1
p=pins[i]
if pin[p]!=pst[i]
then
zn=zb+i
if zn<10
then
=>publish Security/Zone0%0int(zn)% %0int(pin[p])%
else
=>publish Security/Zone%0int(zn)% %0int(pin[p])%
endif
pst[i]=pin[p]
endif
next
if cfg==0
then
=#pubcfg("FrontDoor" "09" "door" "FD")
=#pubcfg("Bedroom"+wind "14" wind "BW")
cfg=1
endif
else
mm = mm + 1
if mm > 30
then
->restart 1
endif
endif

#pubcfg(nam zstr cls uid)
=>publish homeassistant/binary_sensor/sec-zone%zstr%/config {"name":"%nam%", "stat_t":"Security/Zone%zstr%","pl_on":1,"pl_off":0,"dev_cla":"%cls%","uniq_id":"%uid%"}

**Comment by barbudor** [comment]
## [9.2.0.5]
### Changed
- ESP32 increase number of relay GPIOs from 8 to 28

**Reply by bartek666666** [reply]
nice :D

**Comment by arendst** [comment]
## [9.2.0.6]
### Changed
- Increase number of switch GPIOs from 8 to 28
- Increase number of button GPIOs from 4 to 8
